rasa/shared/core/events.py

```diff
@@ -377,27 +409,31 @@ def _from_parse_data(
         )
 
     def __hash__(self) -> int:
+        """Returns unique hash of object."""
         return hash((self.text, self.intent_name, jsonpickle.encode(self.entities)))
 
     @property
     def intent_name(self) -> Optional[Text]:
+        """Returns intent name or `None` if no intent."""
         return self.intent.get(INTENT_NAME_KEY)
 
     def __eq__(self, other: Any) -> bool:
+        """Compares object with other object."""
         if not isinstance(other, UserUttered):
-            return False
-        else:
-            return (
-                self.text,
-                self.intent_name,
-                [jsonpickle.encode(ent) for ent in self.entities],
-            ) == (
-                other.text,
-                other.intent_name,
-                [jsonpickle.encode(ent) for ent in other.entities],
-            )
+            return NotImplemented
+
+        return (
+            self.text,
+            self.intent_name,
+            [jsonpickle.encode(ent) for ent in self.entities],
+        ) == (
+            other.text,
+            other.intent_name,
+            [jsonpickle.encode(ent) for ent in other.entities],
+        )
 
     def __str__(self) -> Text:
+        """Returns text representation of event."""
         return (
             f"UserUttered(text: {self.text}, intent: {self.intent}, "
```

Comment:
I think it should be
```suggestion
            f"UserUttered(text: {self.text}, intent: {self.intent_name}, "
```
otherwise you get all the confidences etc

Reply:
I agree. We should revisit this on `master` once we decided on a Python code convention. Let's focus on getting `e2e` merged and not to fix all problems in the code.

Reply:
what do you mean? I don't think its convention problem. If you use `str` as hash method, you'd get different events all the time, because confidences will be different

Reply:
We don't use `str` as `hash` here. The `hash` actually uses the `intent_name`. `__hash__` is currently not used for hashing on predictions / only on training data.

Reply:
then `hash` is wrong, because it contains `self.entities` which contain confidences and extractors etc, which I don't think should be a part of the hash

Reply:
fixed representation and hashing